### PR TITLE
Add a central module to handle the metric data

### PR DESCRIPTION
**Pull request by georgf**
Having more conversations on the usage of Redux, it does the job of centralizing the data state but might be more complex than what we need.

As an alternative to the Redux migration #26 i'm testing out moving the metric data handling into a separate module, `MetricData` / metricdata.js.

This centrally handles:
- mapping the metric/channel/version selection to a data request
- async fetching of metric data
- getting summary values out of it

From there we can:
- hold an instance of `MetricData` in `App`
- pass around this data store in props instead of always passing many individual props
- keep metric data handling logic in a central place

### CODE REVIEW

**Comment by georgf** [comment]
John, does this approach make sense to you?

**Comment by georgf** [comment]
Or to you Davor?

File: src/app.jsx [hunk 1]
```diff
@@ -140,11 +61,13 @@ class App extends React.Component {
...
-    var pctChange = (rawChange / this.state.lastMedian) * 100;
-    var roundedChange = pctChange.toFixed(2);
-    this.setState({change: roundedChange});
+  componentDidUpdate(prevProps, prevState) {
+    if (prevState.activeMetric !== this.state.activeMetric ||
+        prevState.activeChannel !== this.state.activeChannel ||
+        prevState.activeVersion !== this.state.activeVersion) {
+      this.dataStore.loadDataFor(this.state.activeMetric, this.state.activeChannel, this.state.activeVersion)
+          .then(() => this.setState({activeData: this.dataStore.active.data}));
+    }
```
Comment (openjck) [inline src/app.jsx:70]: Memoization might make this section a little cleaner:

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization

**Comment by openjck** [comment]
I can't say I'm an expert at state management in React, but this seems fine to me. It's pretty straightforward. @spasovski any other feedback?

File: src/components/summaryview.jsx [hunk 2]
```diff
@@ -3,6 +3,14 @@ import {Grid, Row} from "react-bootstrap";
 
 export class SummaryView extends Component {
   render() {
+    let median = this.props.dataStore.median.toFixed(2);
+    let nfifth = this.props.dataStore.nfifthPercentile.toFixed(2);
+    let change = this.props.dataStore.change.toFixed(2);
+
+    let metric = this.props.dataStore.active.metric;
+    let channel = this.props.dataStore.active.channel;
+    let version = this.props.activeVersion;
```
Comment (spasovski) [inline src/components/summaryview.jsx]: nit: these look like they cold be `const`

File: src/metricdata.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  getMean() {
+    let currentData = this._active.data;
+    let buckets = currentData.map(item => item.start)
+                             .concat([this.getLastBucketUpper()]);
+    let values = currentData.map(item => item.count);
+    let totalHits = 0;
+    let bucketHits = 0;
+    let linearTerm = (buckets[buckets.length - 1] - buckets[buckets.length -2]) / 2;
+    let exponentialFactor = Math.sqrt(buckets[buckets.length - 1] / buckets[buckets.length - 2]);
+    let useLinearBuckets = this.kind === "linear" || this.kind === "flag" || this.kind === "boolean" || this.kind === "enumerated";
```
Comment (spasovski) [inline src/metricdata.js]: could also do `let useLinearBuckets = ['linear', 'flag', 'boolean', 'enumerated'].includes(this.kind);`

File: src/metricdata.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+  updateCachedData() {
+    this.cached.mean = this.getMean().toFixed(2);
+    this.cached.median = this.getPercentile(50).toFixed(2);
+    this.cached.nfifth = this.getPercentile(95).toFixed(2);
+    this.cached.change = this.getChange().toFixed(2);
+  }
+
+  loadDataFor(metric, channel, version) {
+    return fetch("https://mozilla.github.io/mdv2/data/" +
```
Comment (spasovski) [inline src/metricdata.js]: the fetch url could be "`https://mozilla.github.io/mdv2/data/${metric}_${channel}_${version}.json`" with string interpolation

File: src/metricdata.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  get channelOptions() {
+    return [
+      "nightly",
+      "beta",
+      "dev edition",
+      "release"
+    ];
+  }
+
+  get versionOptions() {
```
Comment (spasovski) [inline src/metricdata.js:50]: I haven't seen this `get` syntax before. Very java-ey

**Comment by spasovski** [comment]
Am I wrong in thinking Redux wasn't used here? Not trying to vote for or against...that's a slightly more involved conversation. John's momoization link is good reading and some of the components could be pure.

I don't have greater insight into this project but the logic itself seems fine and I like the trimming of the previously massive component into something that gets its state externally.

**Comment by georgf** [comment]
Thanks John & Davor, looking into it.
Flagging Chris for feedback from the project side.

**Comment by georgf** [comment]
> Am I wrong in thinking Redux wasn't used here? Not trying to vote for or against...that's a slightly more involved conversation.

This is correct. I heard some light feedback that Redux would solve the problem at hand (centralizing data handling) but is probably more complex than required. Thus i'm testing this out as an alternative, more straight-forward solution.

File: src/app.jsx (hunk 1 above)
Comment (georgf) [reply src/app.jsx:70]: This is interesting, thanks.
I may have to digest that a bit and look into it in a follow-up.

File: src/metricdata.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+      version: "62",
+      data: GC_MS_nightly_62,
+    };
+
+    this.cached = {
+      change: "",
+      nfifthPercentile: "",
+      median: "",
+      mean: "",
+      lastMedian: 315,
```
Comment (chutten) [inline src/metricdata.js:20]: That's oddly specific

File: src/metricdata.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    };
+  }
+
+  get active() {
+    return {
+      metric: this._active.metric,
+      channel: this._active.channel,
+      version: this._active.version,
+      data: this._active.data,
```
Comment (chutten) [inline src/metricdata.js:29]: While we're having fun with ES6, this can be rewritten as `return {...this._active};`

File: src/metricdata.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  get nfifthPercentile() {
+    return this.getPercentile(95);
+  }
+
+  get change() {
+    return this.getChange();
+  }
+
+  updateCachedData() {
+    this.cached.mean = this.getMean().toFixed(2);
```
Comment (chutten) [inline src/metricdata.js:71]: if `this.mean` returns full-precision numbers, then I think `this.cached.mean` should be full-precision as well.

File: src/metricdata.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    this.cached.median = this.getPercentile(50).toFixed(2);
+    this.cached.nfifth = this.getPercentile(95).toFixed(2);
+    this.cached.change = this.getChange().toFixed(2);
+  }
+
+  async loadDataFor(metric, channel, version) {
+    const response = await fetch(`https://mozilla.github.io/mdv2/data/${metric}_${channel}_${version}.json`);
```
Comment (chutten) [inline src/metricdata.js]: A relative URL could make testing easier

File: src/metricdata.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    const data = await response.json();
+
+    this._active.metric = metric;
+    this._active.channel = channel;
+    this._active.version = version;
+    this._active.data = data;
```
Comment (chutten) [inline src/metricdata.js]: More ES6 wizardry: `this._active = {...this._active, metric, channel, version, data};`

File: src/metricdata.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+  get nfifthPercentile() {
+    return this.getPercentile(95);
+  }
+
+  get change() {
+    return this.getChange();
+  }
+
+  updateCachedData() {
```
Comment (chutten) [inline src/metricdata.js:70]: We aren't caching data, though, we're caching aggregates?

File: src/metricdata.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+    this._active.metric = metric;
+    this._active.channel = channel;
+    this._active.version = version;
+    this._active.data = data;
+    this.updateCachedData();
+  }
+
+  getMean() {
+    let currentData = this._active.data;
```
Comment (chutten) [inline src/metricdata.js:93]: I'm a proponent of making `const` whatever we can make `const`. So here that'd be `currentData, buckets, values, linearTerm, exponentialFactor, useLinearBuckets, centralX` or so.

Comment (chutten) [reply src/metricdata.js:93]: Ohhh, wait this is copied from erin's impl elsewhere. Nevermind, then.

File: src/metricdata.js [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+      return buckets[0] + 1;
+    }
+
+    /*if (this.state.activeMetric.type === "linear" || this.state.activeMetric.type === "flag" || this.state.activeMetric.type ===
+    "boolean" || this.state.activeMetric.type === "enumerated") {
+      return buckets[buckets.length - 1] + buckets[buckets.length - 1]
+      - buckets[buckets.length -2];
+    }*/
+
+    return  buckets[buckets.length - 1] * buckets[buckets.length - 1] / buckets[buckets.length - 2];
```
Comment (chutten) [inline src/metricdata.js:127]: there's two spaces after `return` here

File: src/metricdata.js [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    /*if (["linear", "flag", "boolean", "enumerated"].includes(this.kind)) {
+      return buckets[percentileBucketIndex] + linearTerm * ratioInBar;
+    } else {*/
+    return buckets[percentileBucketIndex] * Math.pow(exponentialFactor, ratioInBar);
+    //}
+  }
+
+  getChange() {
+    let rawChange = this.getPercentile(50) - this.cached.lastMedian;
+    let pctChange = (rawChange / this.cached.lastMedian) * 100;
```
Comment (chutten) [inline src/metricdata.js:157]: `this.cached.lastMedian` is never written to? Oh, I guess that's a TODO.

File: src/metricdata.js (hunk 6 above)
Comment (georgf) [reply src/metricdata.js:20]: This is copied code, let's get to that later.

File: src/metricdata.js (hunk 9 above)
Comment (georgf) [reply src/metricdata.js]: Let's get to that when the need comes up.

File: src/metricdata.js (hunk 11 above)
Comment (georgf) [reply src/metricdata.js:70]: It's still cached data though?
This caching we may want to revisit soon, currently this is mostly copied over and adopted to make work.

File: src/metricdata.js (hunk 14 above)
Comment (georgf) [reply src/metricdata.js:157]: Yes, the `lastMedian` approach needs revisiting anyway (currently it doesn't show it from the preceding version, but from the last-loaded version).

**Comment by georgf** [comment]
@chutten Please take a look over it. Let's aim for identifying improvements as follow-ups where possible.

One follow-up thing (existing issue) i've identified is that switching to *scalars_devtools_onboarding_is_devtools_user* results (in the summary view) in the median value of infinity.
This seems to come from `getLastBucketUpper()` doing a div by `0` as `buckets[buckets.length - 2]` is `0`.
This is not a new issue, it just is visible now because we're actively loading different data files now (we didn't before).

**Review by chutten — APPROVED** [review]
I leave you to decide which comments are appropriate to address now and which as follow-ups (and which to ignore and evaluate later as part of a generic cleanup sweep).

`getLastBucketUpper` has commented out code that requires information about what the type of measurement is. We don't have that information, but `scalars_devtools_onboarding_is_devtools_user` is a metric that requires the special handling. I'm not worried about that at this state as it will require broader changes to make correct.

In general I'm happy with the move and the reorganization, so in my opinion it is merge-able.

**Comment by georgf** [comment]
Ok, moving forward with the merge.
Let's plan a review of the code base to identify what we want to improve.

**Pull request merged**